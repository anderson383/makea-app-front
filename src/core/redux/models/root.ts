import { IProductState } from "./products";

export interface IrootState {
  product: IProductState
}