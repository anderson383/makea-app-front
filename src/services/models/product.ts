import { ICategory } from "./category";

export interface IProduct {
  id: string;
  name: string;
  image: string;
  price: number;
  description: string;
  isAddCart?: boolean;
  categoria?: string | ICategory;
}