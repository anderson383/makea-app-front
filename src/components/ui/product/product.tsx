import { View, Text, StyleSheet, FlatList, ScrollView, Image, Button, TouchableNativeFeedback, TouchableOpacity, Touchable } from "react-native"
import { THEME } from "./../../../theme";
import { SvgXml } from 'react-native-svg';
import { IProduct } from "../../../services/models/product";
import { numberFormat } from "../../../helpers/numberCurrency";
import {HeartIcon, HeartIconActive} from "../../../../assets/icons/heart";

export interface ProductProps {
  product: IProduct
  onClickHeart?: () => void;
  onClickCard?: () => void;
}

const Product:React.FC<ProductProps> = ({product, onClickHeart, onClickCard}) => {
  return (
      <View style={styles.product} key={product.id}>
        <TouchableOpacity onPress={onClickCard && onClickCard} testID="productCard">
          <View style={{flex: 1}}>
            <Image source={{ uri: product.image, height: 150 }} style={{width: '100%'}} />
          </View>
          <View style={styles.info} >
            <View style={styles.priceContent}>
              <Text style={styles.price}>{numberFormat(product.price)}</Text>
              <TouchableOpacity  onPress={onClickHeart && onClickHeart} testID="addCart">
                { product.isAddCart ?  <SvgXml  xml={HeartIconActive()}  /> : <SvgXml  xml={HeartIcon()}  />} 
              </TouchableOpacity>
            </View>
            <View>
            <Text>{product.name}</Text>
            </View>
          </View>
        </TouchableOpacity>
      </View>
   
  )
}

const styles = StyleSheet.create({
  product: {
    marginRight: 3,
    borderRadius: 10,
    flexBasis: '47.8%',
    overflow: 'hidden',
    borderColor: '#f1f1f1',
    borderWidth: 1
  },
  info: {
    padding: 8,
    position: 'relative',
    gap: THEME.paddings.sm
  },
  priceContent: {
    flexDirection: 'row',
    justifyContent: 'space-between'
  },
  price: {
    color: THEME.colors.secondaryColor,
    fontWeight: 'bold'
  }
})

export default Product;