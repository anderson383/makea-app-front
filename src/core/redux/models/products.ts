import { IProduct } from '../../../services/models/product';

export interface IProductState {
  products: IProduct[]
}