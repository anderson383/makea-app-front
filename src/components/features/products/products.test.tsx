
import Products from '.'
import { act } from 'react-dom/test-utils';
import {render, screen, fireEvent, waitFor} from '@testing-library/react-native'
import configureStore from 'redux-mock-store';
import { Provider } from 'react-redux';
import { IProduct } from '@/src/services/models/product';


const navigate = jest.fn()
const moduleRouter = {
  useNavigate: () => navigate
}
jest.mock('react-router-native', () => moduleRouter)

const mockProduct = {
  id: 'idtest',
  name: 'Product test',
  description: '',
  image: '',
  price: 1000,
  categoria: 'categoriatest',
  isAddCart: false
}

jest.mock('../../../services/config/context', () => ({
  useRepositoryIoc: jest.fn().mockReturnValue({
    container: {
      get: jest.fn().mockReturnValue({
        list: jest.fn().mockImplementation(() => Promise.resolve<IProduct[]>([mockProduct]))
      })
    }
  })
}))


const mockStore = configureStore([]);

describe('Test feature products list', () => {
  
  const setup = async (store:any) => {
    return await waitFor(() => render(
      <Provider store={store}>
        <Products title='Page test' code='codetest'/>
      </Provider>
    )) 
  }

  beforeEach(() => {
    jest.clearAllMocks();
  });

  test(`Should render component `, async () => {
    const store = mockStore({
      product: {
        products: []
      }
    });
    await act(async () => {
      const {toJSON} = await setup(store)

      expect(toJSON()).toMatchSnapshot()
    });
  })

  test(`Should render component if products are in the store`, async () => {
    const store = mockStore({
      product: {
        products: [ {...mockProduct, isAddCart: true} ]
      }
    });
    await act(async () => {
      const {toJSON} = await setup(store)

      expect(toJSON()).toMatchSnapshot()
    });
  })


  test(`Should click card for to detail product`, async () => {
    await act(async () => {
      const store = mockStore({
        product: {
          products: []
        }
      });
      const {getByTestId} = await setup(store)

      const button = getByTestId('productCard')
      fireEvent.press(button)
      expect(navigate).toHaveBeenCalledWith('/detail/' + mockProduct.id)
    });
  })

  test(`Should click add to cart`, async () => {
    await act(async () => {
      const store = mockStore({
        product: {
          products: []
        }
      });
      const {getByTestId} = await setup(store)

      const button = getByTestId('addCart')
      fireEvent.press(button)
      
      expect(store.getActions()[0].payload.isAddCart).toBe(true)
    });
  })
})