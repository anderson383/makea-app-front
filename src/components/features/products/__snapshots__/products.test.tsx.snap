// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Test feature products list Should render component  1`] = `
<View
  style={
    {
      "flex": 1,
      "paddingTop": 54,
    }
  }
>
  <View
    style={
      {
        "alignItems": "center",
        "backgroundColor": "#fff",
        "flexDirection": "row",
        "height": 73,
        "justifyContent": "space-between",
        "paddingHorizontal": 16,
      }
    }
  >
    <Text
      style={
        {
          "fontSize": 35,
          "fontWeight": "bold",
        }
      }
    >
      Page test
    </Text>
    <RNSVGSvgView
      align="xMidYMid"
      bbHeight="20"
      bbWidth="20"
      fill="none"
      focusable={false}
      height={20}
      meetOrSlice={0}
      minX={0}
      minY={0}
      style={
        [
          {
            "backgroundColor": "transparent",
            "borderWidth": 0,
          },
          {
            "flex": 0,
            "height": 20,
            "width": 20,
          },
        ]
      }
      vbHeight={20}
      vbWidth={20}
      width={20}
      xml=" 
<svg width="20" height="20" viewBox="0 0 20 20" fill="none" xmlns="http://www.w3.org/2000/svg">
<path d="M13.2678 12.8521C14.5265 11.4896 15.3117 9.66507 15.3117 7.65584C15.3117 3.44108 11.8821 0 7.65584 0C3.42954 0 0 3.44108 0 7.65584C0 11.8706 3.42954 15.3117 7.65584 15.3117C9.3764 15.3117 10.9584 14.7459 12.2401 13.7874L17.9329 19.4802C18.0715 19.6188 18.2447 19.6881 18.4179 19.6881C18.5911 19.6881 18.7758 19.6188 18.9029 19.4802C19.1685 19.2146 19.1685 18.7759 18.9029 18.4987L13.2678 12.8521ZM1.38567 7.65584C1.38567 4.2032 4.19166 1.38567 7.65584 1.38567C11.1085 1.38567 13.926 4.19166 13.926 7.65584C13.926 11.12 11.12 13.926 7.65584 13.926C4.19166 13.926 1.38567 11.1085 1.38567 7.65584Z" fill="#212224"/>
</svg>

"
      xmlns="http://www.w3.org/2000/svg"
    >
      <RNSVGGroup
        fill={null}
        propList={
          [
            "fill",
          ]
        }
      >
        <RNSVGPath
          d="M13.2678 12.8521C14.5265 11.4896 15.3117 9.66507 15.3117 7.65584C15.3117 3.44108 11.8821 0 7.65584 0C3.42954 0 0 3.44108 0 7.65584C0 11.8706 3.42954 15.3117 7.65584 15.3117C9.3764 15.3117 10.9584 14.7459 12.2401 13.7874L17.9329 19.4802C18.0715 19.6188 18.2447 19.6881 18.4179 19.6881C18.5911 19.6881 18.7758 19.6188 18.9029 19.4802C19.1685 19.2146 19.1685 18.7759 18.9029 18.4987L13.2678 12.8521ZM1.38567 7.65584C1.38567 4.2032 4.19166 1.38567 7.65584 1.38567C11.1085 1.38567 13.926 4.19166 13.926 7.65584C13.926 11.12 11.12 13.926 7.65584 13.926C4.19166 13.926 1.38567 11.1085 1.38567 7.65584Z"
          fill={
            {
              "payload": 4280361508,
              "type": 0,
            }
          }
          propList={
            [
              "fill",
            ]
          }
        />
      </RNSVGGroup>
    </RNSVGSvgView>
  </View>
  <RCTScrollView>
    <View>
      <View
        style={
          {
            "flex": 1,
            "flexDirection": "row",
            "flexWrap": "wrap",
            "gap": 8,
            "paddingHorizontal": 16,
          }
        }
      >
        <View
          style={
            {
              "borderColor": "#f1f1f1",
              "borderRadius": 10,
              "borderWidth": 1,
              "marginRight": 3,
              "maxWidth": 172,
              "overflow": "hidden",
              "width": "100%",
            }
          }
        >
          <View
            accessibilityState={
              {
                "busy": undefined,
                "checked": undefined,
                "disabled": undefined,
                "expanded": undefined,
                "selected": undefined,
              }
            }
            accessibilityValue={
              {
                "max": undefined,
                "min": undefined,
                "now": undefined,
                "text": undefined,
              }
            }
            accessible={true}
            collapsable={false}
            focusable={true}
            onClick={[Function]}
            onResponderGrant={[Function]}
            onResponderMove={[Function]}
            onResponderRelease={[Function]}
            onResponderTerminate={[Function]}
            onResponderTerminationRequest={[Function]}
            onStartShouldSetResponder={[Function]}
            style={
              {
                "opacity": 1,
              }
            }
            testID="productCard"
          >
            <View
              style={
                {
                  "flex": 1,
                }
              }
            >
              <Image
                source={
                  {
                    "height": 150,
                    "uri": "",
                  }
                }
                style={
                  {
                    "width": "100%",
                  }
                }
              />
            </View>
            <View
              style={
                {
                  "gap": 8,
                  "padding": 8,
                  "position": "relative",
                }
              }
            >
              <View
                style={
                  {
                    "flexDirection": "row",
                    "justifyContent": "space-between",
                  }
                }
              >
                <Text
                  style={
                    {
                      "color": "#FE2550",
                      "fontWeight": "bold",
                    }
                  }
                >
                  $Â 1,000
                </Text>
                <View
                  accessibilityState={
                    {
                      "busy": undefined,
                      "checked": undefined,
                      "disabled": undefined,
                      "expanded": undefined,
                      "selected": undefined,
                    }
                  }
                  accessibilityValue={
                    {
                      "max": undefined,
                      "min": undefined,
                      "now": undefined,
                      "text": undefined,
                    }
                  }
                  accessible={true}
                  collapsable={false}
                  focusable={true}
                  onClick={[Function]}
                  onResponderGrant={[Function]}
                  onResponderMove={[Function]}
                  onResponderRelease={[Function]}
                  onResponderTerminate={[Function]}
                  onResponderTerminationRequest={[Function]}
                  onStartShouldSetResponder={[Function]}
                  style={
                    {
                      "opacity": 1,
                    }
                  }
                  testID="addCart"
                >
                  <RNSVGSvgView
                    align="xMidYMid"
                    bbHeight="16"
                    bbWidth="18"
                    fill="none"
                    focusable={false}
                    height={16}
                    meetOrSlice={0}
                    minX={0}
                    minY={0}
                    style={
                      [
                        {
                          "backgroundColor": "transparent",
                          "borderWidth": 0,
                        },
                        {
                          "flex": 0,
                          "height": 16,
                          "width": 18,
                        },
                      ]
                    }
                    vbHeight={16}
                    vbWidth={18}
                    width={18}
                    xml=" 
<svg width="18" height="16" viewBox="0 0 18 16" fill="none" xmlns="http://www.w3.org/2000/svg">
<path fill-rule="evenodd" clip-rule="evenodd" d="M9 1.50572C9.983 0.539931 11.2784 0 12.6634 0C15.606 0 18 2.47813 18 5.52414C18 11.1957 9.6866 15.6307 9.3326 15.8168L8.984 16L8.6424 15.8029C8.2896 15.5994 0 10.7572 0 5.52414C0 2.47813 2.394 0 5.3366 0C6.7214 0 8.0166 0.540138 9 1.50572ZM1.4694 5.52415C1.4694 8.95668 6.6198 12.8107 9.0114 14.2686C10.5278 13.3979 16.5306 9.68646 16.5306 5.52415C16.5306 3.31681 14.7958 1.52104 12.6634 1.52104C11.4458 1.52104 10.3234 2.09741 9.5846 3.10212L9 3.8969L8.4154 3.10212C7.6764 2.09741 6.5542 1.52104 5.3366 1.52104C3.2042 1.52104 1.4694 3.31681 1.4694 5.52415Z" fill="#212224"/>
</svg>
"
                    xmlns="http://www.w3.org/2000/svg"
                  >
                    <RNSVGGroup
                      fill={null}
                      propList={
                        [
                          "fill",
                        ]
                      }
                    >
                      <RNSVGPath
                        clipRule={0}
                        d="M9 1.50572C9.983 0.539931 11.2784 0 12.6634 0C15.606 0 18 2.47813 18 5.52414C18 11.1957 9.6866 15.6307 9.3326 15.8168L8.984 16L8.6424 15.8029C8.2896 15.5994 0 10.7572 0 5.52414C0 2.47813 2.394 0 5.3366 0C6.7214 0 8.0166 0.540138 9 1.50572ZM1.4694 5.52415C1.4694 8.95668 6.6198 12.8107 9.0114 14.2686C10.5278 13.3979 16.5306 9.68646 16.5306 5.52415C16.5306 3.31681 14.7958 1.52104 12.6634 1.52104C11.4458 1.52104 10.3234 2.09741 9.5846 3.10212L9 3.8969L8.4154 3.10212C7.6764 2.09741 6.5542 1.52104 5.3366 1.52104C3.2042 1.52104 1.4694 3.31681 1.4694 5.52415Z"
                        fill={
                          {
                            "payload": 4280361508,
                            "type": 0,
                          }
                        }
                        fillRule={0}
                        propList={
                          [
                            "fill",
                            "fillRule",
                          ]
                        }
                      />
                    </RNSVGGroup>
                  </RNSVGSvgView>
                </View>
              </View>
              <View>
                <Text>
                  Product test
                </Text>
              </View>
            </View>
          </View>
        </View>
      </View>
    </View>
  </RCTScrollView>
</View>
`;

exports[`Test feature products list Should render component if products are in the store 1`] = `
<View
  style={
    {
      "flex": 1,
      "paddingTop": 54,
    }
  }
>
  <View
    style={
      {
        "alignItems": "center",
        "backgroundColor": "#fff",
        "flexDirection": "row",
        "height": 73,
        "justifyContent": "space-between",
        "paddingHorizontal": 16,
      }
    }
  >
    <Text
      style={
        {
          "fontSize": 35,
          "fontWeight": "bold",
        }
      }
    >
      Page test
    </Text>
    <RNSVGSvgView
      align="xMidYMid"
      bbHeight="20"
      bbWidth="20"
      fill="none"
      focusable={false}
      height={20}
      meetOrSlice={0}
      minX={0}
      minY={0}
      style={
        [
          {
            "backgroundColor": "transparent",
            "borderWidth": 0,
          },
          {
            "flex": 0,
            "height": 20,
            "width": 20,
          },
        ]
      }
      vbHeight={20}
      vbWidth={20}
      width={20}
      xml=" 
<svg width="20" height="20" viewBox="0 0 20 20" fill="none" xmlns="http://www.w3.org/2000/svg">
<path d="M13.2678 12.8521C14.5265 11.4896 15.3117 9.66507 15.3117 7.65584C15.3117 3.44108 11.8821 0 7.65584 0C3.42954 0 0 3.44108 0 7.65584C0 11.8706 3.42954 15.3117 7.65584 15.3117C9.3764 15.3117 10.9584 14.7459 12.2401 13.7874L17.9329 19.4802C18.0715 19.6188 18.2447 19.6881 18.4179 19.6881C18.5911 19.6881 18.7758 19.6188 18.9029 19.4802C19.1685 19.2146 19.1685 18.7759 18.9029 18.4987L13.2678 12.8521ZM1.38567 7.65584C1.38567 4.2032 4.19166 1.38567 7.65584 1.38567C11.1085 1.38567 13.926 4.19166 13.926 7.65584C13.926 11.12 11.12 13.926 7.65584 13.926C4.19166 13.926 1.38567 11.1085 1.38567 7.65584Z" fill="#212224"/>
</svg>

"
      xmlns="http://www.w3.org/2000/svg"
    >
      <RNSVGGroup
        fill={null}
        propList={
          [
            "fill",
          ]
        }
      >
        <RNSVGPath
          d="M13.2678 12.8521C14.5265 11.4896 15.3117 9.66507 15.3117 7.65584C15.3117 3.44108 11.8821 0 7.65584 0C3.42954 0 0 3.44108 0 7.65584C0 11.8706 3.42954 15.3117 7.65584 15.3117C9.3764 15.3117 10.9584 14.7459 12.2401 13.7874L17.9329 19.4802C18.0715 19.6188 18.2447 19.6881 18.4179 19.6881C18.5911 19.6881 18.7758 19.6188 18.9029 19.4802C19.1685 19.2146 19.1685 18.7759 18.9029 18.4987L13.2678 12.8521ZM1.38567 7.65584C1.38567 4.2032 4.19166 1.38567 7.65584 1.38567C11.1085 1.38567 13.926 4.19166 13.926 7.65584C13.926 11.12 11.12 13.926 7.65584 13.926C4.19166 13.926 1.38567 11.1085 1.38567 7.65584Z"
          fill={
            {
              "payload": 4280361508,
              "type": 0,
            }
          }
          propList={
            [
              "fill",
            ]
          }
        />
      </RNSVGGroup>
    </RNSVGSvgView>
  </View>
  <RCTScrollView>
    <View>
      <View
        style={
          {
            "flex": 1,
            "flexDirection": "row",
            "flexWrap": "wrap",
            "gap": 8,
            "paddingHorizontal": 16,
          }
        }
      >
        <View
          style={
            {
              "borderColor": "#f1f1f1",
              "borderRadius": 10,
              "borderWidth": 1,
              "marginRight": 3,
              "maxWidth": 172,
              "overflow": "hidden",
              "width": "100%",
            }
          }
        >
          <View
            accessibilityState={
              {
                "busy": undefined,
                "checked": undefined,
                "disabled": undefined,
                "expanded": undefined,
                "selected": undefined,
              }
            }
            accessibilityValue={
              {
                "max": undefined,
                "min": undefined,
                "now": undefined,
                "text": undefined,
              }
            }
            accessible={true}
            collapsable={false}
            focusable={true}
            onClick={[Function]}
            onResponderGrant={[Function]}
            onResponderMove={[Function]}
            onResponderRelease={[Function]}
            onResponderTerminate={[Function]}
            onResponderTerminationRequest={[Function]}
            onStartShouldSetResponder={[Function]}
            style={
              {
                "opacity": 1,
              }
            }
            testID="productCard"
          >
            <View
              style={
                {
                  "flex": 1,
                }
              }
            >
              <Image
                source={
                  {
                    "height": 150,
                    "uri": "",
                  }
                }
                style={
                  {
                    "width": "100%",
                  }
                }
              />
            </View>
            <View
              style={
                {
                  "gap": 8,
                  "padding": 8,
                  "position": "relative",
                }
              }
            >
              <View
                style={
                  {
                    "flexDirection": "row",
                    "justifyContent": "space-between",
                  }
                }
              >
                <Text
                  style={
                    {
                      "color": "#FE2550",
                      "fontWeight": "bold",
                    }
                  }
                >
                  $Â 1,000
                </Text>
                <View
                  accessibilityState={
                    {
                      "busy": undefined,
                      "checked": undefined,
                      "disabled": undefined,
                      "expanded": undefined,
                      "selected": undefined,
                    }
                  }
                  accessibilityValue={
                    {
                      "max": undefined,
                      "min": undefined,
                      "now": undefined,
                      "text": undefined,
                    }
                  }
                  accessible={true}
                  collapsable={false}
                  focusable={true}
                  onClick={[Function]}
                  onResponderGrant={[Function]}
                  onResponderMove={[Function]}
                  onResponderRelease={[Function]}
                  onResponderTerminate={[Function]}
                  onResponderTerminationRequest={[Function]}
                  onStartShouldSetResponder={[Function]}
                  style={
                    {
                      "opacity": 1,
                    }
                  }
                  testID="addCart"
                >
                  <RNSVGSvgView
                    align="xMidYMid"
                    bbHeight="16"
                    bbWidth="18"
                    fill="none"
                    focusable={false}
                    height={16}
                    meetOrSlice={0}
                    minX={0}
                    minY={0}
                    style={
                      [
                        {
                          "backgroundColor": "transparent",
                          "borderWidth": 0,
                        },
                        {
                          "flex": 0,
                          "height": 16,
                          "width": 18,
                        },
                      ]
                    }
                    vbHeight={16}
                    vbWidth={18}
                    width={18}
                    xml=" 
<svg width="18" height="16" viewBox="0 0 18 16" fill="none" xmlns="http://www.w3.org/2000/svg">
<path d="M12.6634 0C11.2784 0 9.983 0.539931 9 1.50572C8.0166 0.540138 6.7214 0 5.3366 0C2.394 0 0 2.47813 0 5.52414C0 10.7572 8.2896 15.5994 8.6424 15.8029L8.984 16L9.3326 15.8168C9.6866 15.6307 18 11.1957 18 5.52414C18 2.47813 15.606 0 12.6634 0Z" fill="#FE2550"/>
</svg>

"
                    xmlns="http://www.w3.org/2000/svg"
                  >
                    <RNSVGGroup
                      fill={null}
                      propList={
                        [
                          "fill",
                        ]
                      }
                    >
                      <RNSVGPath
                        d="M12.6634 0C11.2784 0 9.983 0.539931 9 1.50572C8.0166 0.540138 6.7214 0 5.3366 0C2.394 0 0 2.47813 0 5.52414C0 10.7572 8.2896 15.5994 8.6424 15.8029L8.984 16L9.3326 15.8168C9.6866 15.6307 18 11.1957 18 5.52414C18 2.47813 15.606 0 12.6634 0Z"
                        fill={
                          {
                            "payload": 4294845776,
                            "type": 0,
                          }
                        }
                        propList={
                          [
                            "fill",
                          ]
                        }
                      />
                    </RNSVGGroup>
                  </RNSVGSvgView>
                </View>
              </View>
              <View>
                <Text>
                  Product test
                </Text>
              </View>
            </View>
          </View>
        </View>
      </View>
    </View>
  </RCTScrollView>
</View>
`;
